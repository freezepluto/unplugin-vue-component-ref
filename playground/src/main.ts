import { createApp } from 'vue';
import App from './app';
import unpluginVueRef from '../../client_/index';

createApp(App).use(unpluginVueRef);
