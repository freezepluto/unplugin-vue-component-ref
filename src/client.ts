import { App, VNodeProps, VNode } from 'vue';

const onVnodeBeforeMountRef_: VNodeProps['onVnodeBeforeMount'] = (
	VNode: VNode
) => {
	const { component } = VNode;
	if (component) {
		component.exposeProxy = new Proxy(
			{},
			{
				get(_, key: string) {
					if (component.exposed && component.exposed[key])
						return component.exposed[key];
					//@ts-ignore
					return component.proxy[key];
				}
			}
		);
	}
};

export default (app: App) => {
	app.config.globalProperties.onVnodeBeforeMountRef_ = onVnodeBeforeMountRef_;
};
