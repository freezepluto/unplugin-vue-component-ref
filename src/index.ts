import { compileTemplate, parse, MagicString } from '@vue/compiler-sfc';
import { findProp } from '@vue/compiler-core';
import type { NodeTransform } from '@vue/compiler-core';
import { onVnodeBeforeMountRef } from './helper';

export const transform = (source: string, filename: string) => {
	const { descriptor } = parse(source);
	const magicString = new MagicString(source);
	const onRefNode: NodeTransform = (node) => {
		if (node.type === 1 /** NodeTypes.ELEMENT */) {
			const ref = findProp(node, 'ref');
			if (ref && ref.type === 6 /**  NodeTypes.ATTRIBUTE */) {
				const { loc } = ref;
				magicString.appendLeft(loc.end.offset, onVnodeBeforeMountRef);
			}
		}
	};
	compileTemplate({
		filename,
		source: descriptor.source,
		id: filename,
		compilerOptions: {
			nodeTransforms: [onRefNode]
		}
	});
	return magicString.toString();
};
