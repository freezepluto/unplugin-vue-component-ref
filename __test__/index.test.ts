import { transform } from '../src/index';
import { test, expect } from 'vitest';

test('setup ref', () => {
	const result = transform(
		`<template>
     <div ref='xx'>5</div>
    </template>
  <script setup>
  let a= 10;
  </script>`,
		'setup'
	);
	expect(result).toMatchSnapshot();
});

test('defineComponent ref', () => {
	const result = transform(
		`<template>
    <div ref='aa'>5</div>
  </template>
  <script >
  import {defineComponent} from "vue"
  export default defineComponent({})
  </script>`,
		'setup'
	);
	expect(result).toMatchSnapshot();
});

test('multiple ref', () => {
	const result = transform(
		`<template>
    <div ref='div1' >
      <div ref="div2">
          <Home ref="home"class="test"></Home>
      </div>
    </div>
  </template>
  <script >
  import {defineComponent} from "vue"
  export default defineComponent({})
  </script>`,
		'setup'
	);
	expect(result).toMatchSnapshot();
});
