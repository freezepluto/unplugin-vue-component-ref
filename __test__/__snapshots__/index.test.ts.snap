// Vitest Snapshot v1

exports[`defineComponent ref 1`] = `
"<template>
    <div ref='aa' :onVnodeBeforeMount=\\"onVnodeBeforeMountRef_\\" >5</div>
  </template>
  <script >
  import {defineComponent} from \\"vue\\"
  export default defineComponent({})
  </script>"
`;

exports[`multiple ref 1`] = `
"<template>
    <div ref='div1' :onVnodeBeforeMount=\\"onVnodeBeforeMountRef_\\"  >
      <div ref=\\"div2\\" :onVnodeBeforeMount=\\"onVnodeBeforeMountRef_\\" >
          <Home ref=\\"home\\" :onVnodeBeforeMount=\\"onVnodeBeforeMountRef_\\" class=\\"test\\"></Home>
      </div>
    </div>
  </template>
  <script >
  import {defineComponent} from \\"vue\\"
  export default defineComponent({})
  </script>"
`;

exports[`setup ref 1`] = `
"<template>
     <div ref='xx' :onVnodeBeforeMount=\\"onVnodeBeforeMountRef_\\" >5</div>
    </template>
  <script setup>
  let a= 10;
  </script>"
`;
